Round3[Step1] - Jay

shin-jae-uk opened the pull request:
@Gundy93 

안녕하세요, 건디! 
Round3 Step1 PR 드립니다.

- 구조체 생성(`Person`, `CoffeeShop`)
  - `class`가 아닌 `struct`로 선언한 이유: 
    - 타입 내부의 값을 `참조`하여 전달할 경우는 없어 보임
    - 현재 단계에서 `상속` 기능이 필요하지 않아 보임
    - 프로퍼티에 굳이 기본값을 지정할 필요가 없어 보임(`memberwise`)
    - Swift는 구조체와 열거형 사용을 선호한다고 들어서😅

- `Person` 구조체
  - `buyCoffee` 메서드: `typesOfCoffee` 열거형 타입의 값을 인자로 받아 문자열 출력

- `CoffeeShop` 구조체
  - `makeCoffee` 메서드: 
    - 자신의 프로퍼티의 값을 변경하므로 `mutating` 키워드 입력
    - `typesOfCoffee`의 값을 받아 `pickUpTable` 배열에 추가
 ---
- 고민했던 점
  - 타입 내부에서 변수와 상수의 결정 (컴파일 오류가 없다면 되도록 상수로 선언했습니다)
  - 구조체와 클래스의 결정 
  값 타입이고 ARC를 사용하지 않아서..? 비교적 가볍다는 구조체의 장점이 있긴 하지만, 
  참조와 상속 등 입체적인 코딩이 필요하다면 클래스가 보다 상위 호환의 개념 같다는 생각을 했습니다.

---
검토 부탁드립니다.
고맙습니다🙏

shin-jae-uk committed:
Step1 - 타입과 인스턴스의 관계

shin-jae-uk committed:
Update some details

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
 //
 
-import Foundation
+struct Person {
+    let name: String
+    var money: Int
+    
+    func buyCoffee(coffee: typesOfCoffee) {
+        print("\(name)이(가) \(coffee.rawValue)를 주문합니다.")
+    }
+}
```
타입의 정의부는 따로 파일을 만들어서 관리하는 것이 좋습니다. 가독성이나 유지보수적인 측면에서 말이죠. 특정 타입과 관련된 코드는 그 파일을 찾아가면 되겠거니 할테니까요.

보통 타입의 이름을 딴 파일로 타입을 관리합니다. 이 경우 `Person.swift`가 되겠네요. 😃

shin-jae-uk replied:
- 타입의 정의는 swift 파일을 구분하여 관리하는군요!
각 타입에 대해 별도로 분리했습니다. (`main`, `Person`, `CoffeeShop`, `Coffee`)

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+}
 
-print("Hello, World!")
+struct CoffeeShop {
+    var totalSales: Int = 0
+    var menu: [typesOfCoffee: Int] = [:]
+    var pickUpTable: [String] = []
+    let barista: Person
+    
+    func takeOrders() {
+        print("바리스타 \(barista.name)이(가) 주문을 받습니다.")
+    }
+    
+    mutating func makeCoffee(coffee: typesOfCoffee) {
+        pickUpTable.append(coffee.rawValue)
+        print("주문받은 \(coffee.rawValue)를 만들었습니다.")
+    }
+}
 
+enum typesOfCoffee: String {
+    case espresso = "에스프레소"
+    case americano = "아메리카노"
+    case iceAmericano = "아이스 아메리카노"
+    case cafeLatte = "카페라떼"
+    case cafeMocha = "카페모카"
+    case vanillaLatte = "바닐라라떼"
+}
```
이 `CoffeeShop`과 `typesOfCoffee` 또한 따로 파일을 만들어 관리하고, `main.swift`에는 실행시키기 위한 코드만 작성되어 있는 것이 좋겠습니다.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+    let name: String
+    var money: Int
+    
+    func buyCoffee(coffee: typesOfCoffee) {
+        print("\(name)이(가) \(coffee.rawValue)를 주문합니다.")
+    }
```
이 메서드의 이름은 `buyCoffee`인 반면에 내부에서 출력되는 문장은 "주문"이라는 단어입니다. `order`와 같은 단어가 더 적절할 것 같습니다. 혹은 "구매합니다"와 같이 내부가 변경될 수도 있을 것 같습니다.

그리고 현재 이 메서드를 읽어보면 "buy coffee coffee"가 되어서 동어반복으로 인해 어색함이 생길 것 같습니다.

`buy(coffee: typesOfCoffee)`와 같이 만들면 호출할 때도 `buy(coffee: .americano)`가 되어 "buy coffee americano" 처럼 보다 매끄러운 네이밍이 될 것 같습니다. 혹은 `buyCoffee(_ coffee: typesOfCoffee)`처럼 만들면 호출할 때 `buyCoffee(.americano)`가 되어 이 역시도 자연스럽습니다.

[Omitting Argument Labels](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions#Omitting-Argument-Labels)를 참고해보시면 좋을 것 같습니다.

shin-jae-uk replied:
- 구매와 주문은 엄연히 다른 뜻이겠네요 👀
의미를 바로잡기 위해 `order`로 수정하고, 코드 리딩에 동어 반복되지 않도록 인수 레이블을 생략했습니다.
- 말씀주신 [Omitting Argument Labels](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions#Omitting-Argument-Labels)를 통해 메서드 호출시 인수 레이블을 생략하는 법을 배웠습니다!

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
-print("Hello, World!")
+struct CoffeeShop {
+    var totalSales: Int = 0
+    var menu: [typesOfCoffee: Int] = [:]
```
`menu`가 빈 값으로 초기화되는데, 그 이유는 무엇인가요? 카페라면 보통 메뉴판이 비어있으리라고 생각되지는 않는 것 같습니다.

shin-jae-uk replied:
- 우선은 요구사항에 구현내용이 기재되어있지 않아 빈 값으로 두었습니다.
그리고 해당 구조체를 통해 여러 인스턴스(CoffeeShop A, B, yagombucks 등)를 생성하면, 각 커피숍에서 취급하는 메뉴도 다르고 가격도 다를 수 있을 것이라 생각했습니다.
- 메뉴판이 비어있으리라 생각되지 않는다는 의미가, 아래에서 어떤 수정사항이 필요하다는 것인지 잘 모르겠어요🥲
  1) 기본 값을 입력해주기 ex) `[.espresso: 4000, .americano: 5000]`
  2) `init` 함수를 통해 `menu` 프로퍼티의 기본 값을 초기화하기

Gundy93 replied:
> init 함수를 통해 menu 프로퍼티의 기본 값을 초기화하기

이 쪽이 더 적절할 것 같습니다. 제이님 말대로 카페마다 다른 가격에 서로 다른 메뉴들을 다룰테니까요. 빽다방과 별다방의 가격이 같을 순 없을테니 ㅎ...

빈 값으로 초기화하지 않는다면 그 초기값은 이니셜라이저에서 전달받아야하겠죠. `Person`의 `name`과 `money`가 그러하듯이.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+    var totalSales: Int = 0
+    var menu: [typesOfCoffee: Int] = [:]
+    var pickUpTable: [String] = []
+    let barista: Person
```
`barista`가 `let`이라는 것은 이 바리스타는 근무교대할 수 없다는 의미로 보입니다. 😂

바리스타가 변경될 일이 없다고 생각하시나요?

shin-jae-uk replied:
바리스타가 변경될 수 있으므로 `var`로 수정했습니다.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+        print("바리스타 \(barista.name)이(가) 주문을 받습니다.")
+    }
+    
+    mutating func makeCoffee(coffee: typesOfCoffee) {
```
이 메서드의 이름도 앞서 코멘트한 `buyCoffee`의 사례와 같이 고민해보시면 좋을 것 같습니다.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+    }
+}
 
+enum typesOfCoffee: String {
```
이 열거형의 이름은 `typesOfCoffee`인데요, 단순히 `Coffee`로 괜찮을 것 같습니다. 더군다나 요구사항에도 이름이 `Coffee`로 명시되어 있긴 한데, 혹시 이렇게 네이밍한 이유가 있을까요?

shin-jae-uk replied:
- ‘타입명도 변수나 함수처럼 구체적으로 적으면 좋지 않을까?’ 하는 의도에서 길게 작성했었습니다😂 
- 요구사항과 같이 `Coffee`도 수정했습니다. 
추가로 타입명에 기존 코드에 소문자 카멜케이스를 사용하는 실수가 있었는데, 대문자 카멜케이스로 변경했습니다.

Gundy93 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -6,7 +6,44 @@
+}
 
+enum typesOfCoffee: String {
+    case espresso = "에스프레소"
+    case americano = "아메리카노"
+    case iceAmericano = "아이스 아메리카노"
+    case cafeLatte = "카페라떼"
+    case cafeMocha = "카페모카"
+    case vanillaLatte = "바닐라라떼"
```
다양한 커피를 열거형 케이스로 정의하셨군요. 좋습니다. 👍

다만 아메리카노만 hot과 ice를 구분한 이유가 궁금하긴 하네요. 서로 별개의 커피로 구분하시는 건가요?

shin-jae-uk replied:
- 큰 의미 없이 별개의 메뉴로 구분했었습니다.
- ice 라는 속성(?)은 모든 메뉴에 적용될 수 있으므로 추후 범용성을 고려하여 별도로 빼놓는 것이 좋을 것 같긴 합니다. 
`[ 아이스로 변경: 0 ]` 이런 식으로요!
- 좀 더 나아가 메뉴를 확장한다면 사이즈 변경 가격도 추가할 수 있을 것 같아요. `[Tall 사이즈(기본): 0, Grande 사이즈: 1000, Venti 사이즈: 1500]`

Gundy93 replied:
지금 단계에서 필요하진 않을 수 있는 코멘트입니다. 현재 요구되는 `Coffee`는 종류만 정의하는 열거형이지만 더 세세하가 커피의 특성들을 정의해주고 싶다면 다양한 특성들을 열거형으로 묶고 상위에서 구조체를 사용할 수도 있을 것 같습니다.

```swift
struct Coffee {
    enum Kind {
        case espresso
        case americano
        case cafeLatte
        case cafeMocha
        case vanillaLatte
    }
    
    enum Temperature {
        case hot
        case ice
    }
    
    enum Size {
        case short
        case tall
        case grande
        case venti
    }
    
    let kind: Kind
    let temperature: Temperature
    let size: Size
}

let shortIceAmericano = Coffee(kind: .americano, temperature: .ice, size: .short)
```

물론 꼭 중첩타입이어야 하는 것은 아니지만 외부에서 사용할 일이 없다면 이렇게도 만들 수 있겠다는 하나의 예시입니다. 중첩타입으로 만들면 타입 이름짓기가 편해지는 부분도 있습니다. `Kind` 타입은 밖에서 보면 `Coffee` 타입 안에 있는 `Coffee.Kind`가 되어서 직관적으로 커피의 종류를 의미함을 알 수 있습니다.

Gundy93 reviewed (commented):
안녕하세요 제이님! 이번 스텝도 충실히 구현해오셨군요. 좋습니다. 👍

> class가 아닌 struct로 선언한 이유:
> - 타입 내부의 값을 참조하여 전달할 경우는 없어 보임

이 말은 정확하게는 '해당 타입들의 인스턴스를 참조할 일이 없어 보인다'라고 이해하면 될까요?

> - 현재 단계에서 상속 기능이 필요하지 않아 보임
> - 프로퍼티에 굳이 기본값을 지정할 필요가 없어 보임(memberwise)

클래스 또한 프로퍼티에 기본값을 지정하지 않아도 괜찮습니다.

뒤에 (memberwise)라고 덧붙이신 걸 보면 '이니셜라이저를 따로 만들 필요가 없어보인다'라고 이해하면 될까요? 하지만 그 이유는 클래스와 구조체 선정에서 설득력을 가지진 못 할 것 같습니다. 이니셜라이저야 만들면 되니까요 ㅎㅎ.

혹시 다른 의미라면 또 알려주시길 바랍니다.

> - Swift는 구조체와 열거형 사용을 선호한다고 들어서😅

구조체와 열거형, 즉 값 타입의 사용을 선호하는 이유는 분명히 있습니다. 하지만 그렇게 들었다고 하면 근거로서 적절하지 않은 것 같습니다. 왜 값 타입을 사용하는 것을 선호하는지까지 알아야 좋을 것 같습니다.

[Choosing Between Structures and Classes](https://developer.apple.com/documentation/swift/choosing_between_structures_and_classes) 문서를 읽어보시면 좋을 것 같습니다.

---

> 고민했던 점
> - 타입 내부에서 변수와 상수의 결정 (컴파일 오류가 없다면 되도록 상수로 선언했습니다)

기본적으로 상수를 선택하셨다는 부분이 아주 좋습니다. 변수는 말그대로 변경될 우려가 있기 때문에 해당 변수를 최초에 작성한 사람이 의도하지 않은 변경을 추후에 할 수 있습니다. 이런 상황을 휴먼 에러라고도 표현할 수 있을 것 같네요.

기본적으로 개발은 여러 사람이 같이하기 때문에 자신이 아니라 다른 개발자가 해당 변수를 수정할 수도 있는데요, 변경을 의도하지 않는다면 `let`으로 선언해서 휴먼 에러를 미연에 방지하는 것이 좋습니다. 잘 하고 계시네요. 👍

> - 구조체와 클래스의 결정
값 타입이고 ARC를 사용하지 않아서..? 비교적 가볍다는 구조체의 장점이 있긴 하지만,
참조와 상속 등 입체적인 코딩이 필요하다면 클래스가 보다 상위 호환의 개념 같다는 생각을 했습니다.

ARC는 자동 참조 카운팅이라는 참조 타입의 메모리 할당 및 해제와 관련된 개념입니다. 그러므로 ARC를 사용하지 않는다 보다는 참조를 사용하지 않는다고 표현하시는 것이 더 적절할 것 같아요!

참조는 몰라도 상속을 통한 입체적인, 즉 객체 지향적인 코드가 필요하다면 클래스를 사용할 것이라고 표현하셨는데, Swift의 프로토콜은 다른 언어의 그것들보다 강력하기 때문에 프로토콜과 값타입을 주로 사용하는 프로토콜 지향 프로그래밍을 권장합니다.

구조체 단품만 놓고 보면 클래스보다 기능이 적겠지만, 프로토콜과 함께 사용한다면 더욱 입체적으로 사용할 수도 있을 것 같습니다. 그래서 저는 보통 참조가 필요한 상황이 아니면 구조체를 사용하고, 추상화는 프로토콜로 하는 것을 좋아합니다.

뭐 그런데 당장 그렇게 작성하는 것은 쉽지 않기 때문에 지금은 "일단 구조체를 쓰자" 정도로 좋은 것 같습니다.

shin-jae-uk committed:
chore: Separate swift files by type

shin-jae-uk committed:
refactor: Apply code review feedback

- Correct naming(types, method)
- Use 'Omitting Argument Labels' for methods
- Modify other details

Shin-Jae-Uk commented:
- 제가 아직 개념의 체계화가 부족한데, 이렇게 잘 캐치해주셔서 감사합니다👍
어떤 행동을 한 이유에 대해 정확히 설명할 수 있는 능력도 꾸준히 길러야 할 것 같아요🫠
- [Choosing Between Structures and Classes](https://developer.apple.com/documentation/swift/choosing_between_structures_and_classes)문서를 보니 좀 더 감이 오는 것 같아요.
  - 구조체를 사용하면 코드의 추론이 더욱 쉽다는 점 → 값 유형이기에 로컬의 변동사항은 다른 파트의 코드에 반영되지 않음
  - 클래스는 Objective-C 상호운용성이 필요할 때, data identity의 컨트롤이 필요할 때 사용함
- 프로토콜이라는 개념에 대해서도 새롭게 알아갑니다! (해당 기능에 필요한 요구 사항을 선언해두는것(청사진), 구현이 아닌 선언)
